spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpRequest.java

```diff
@@ -0,0 +1,174 @@
+	}
+
+	@Override
+	public DataBufferFactory bufferFactory() {
+		return this.bufferFactory;
+	}
+
+	@Override
+	public Mono<Void> writeWith(Publisher<? extends DataBuffer> publisher) {
+		String contentType = jettyRequest.getHeaders().contains(HttpHeader.CONTENT_TYPE) ?
+				jettyRequest.getHeaders().getField(HttpHeader.CONTENT_TYPE).getValue():
+				MediaType.APPLICATION_OCTET_STREAM_VALUE;
+		Publisher<ContentChunk> chunks = Flux.from(publisher).map(buffer -> new ContentChunk(buffer.asByteBuffer(), new Callback() {
+			@Override
+			public void succeeded() {
+				DataBufferUtils.release(buffer);
+			}
+
+			@Override
+			public void failed(Throwable x) {
+				DataBufferUtils.release(buffer);
+				throw Exceptions.propagate(x);
+			}
+		}));
+		ReactiveRequest.Content requestContent = ReactiveRequest.Content.fromPublisher(chunks, contentType);
+		this.reactiveRequest = ReactiveRequest.newBuilder(jettyRequest).content(requestContent).build();
+		return doCommit(this::completes);
+	}
+
+	@Override
+	public Mono<Void> writeAndFlushWith(Publisher<? extends Publisher<? extends DataBuffer>> body) {
+		String contentType = this.jettyRequest.getHeaders().getField(HttpHeader.CONTENT_TYPE).getValue();
+		ReactiveRequest.Content content = ReactiveRequest.Content.fromPublisher(Flux
+				.from(body)
+				.flatMap(Function.identity())
+				.map(buffer -> new ContentChunk(buffer.asByteBuffer(), new Callback() {
+					@Override
+					public void succeeded() {
+						DataBufferUtils.release(buffer);
+					}
+
+					@Override
+					public void failed(Throwable x) {
+						DataBufferUtils.release(buffer);
+						throw Exceptions.propagate(x);
+					}
+				})), contentType);
+		this.reactiveRequest = ReactiveRequest.newBuilder(this.jettyRequest).content(content).build();
+		return doCommit(this::completes);
+	}
+
+	@Override
+	public Mono<Void> setComplete() {
+		if (this.reactiveRequest == null) {
+			this.reactiveRequest = ReactiveRequest.newBuilder(this.jettyRequest).build();
+		}
+		return doCommit(this::completes);
+	}
+
+	private Mono<Void> completes() {
+		this.response = Mono.from(
+				this.reactiveRequest.response((reactiveResponse, contentChunks) -> {
+
+					// Implementation with buffer copy instead of optimized buffer wrapping
+					// because the latter hangs since Callback#succeeded doesn't allow
+					// releasing the buffer and requesting more data at different times
+					// (required for Mono<DataBuffer> for example).
+					// See https://github.com/eclipse/jetty.project/issues/2429
+					Flux<DataBuffer> content = Flux.from(contentChunks).map(chunk -> {
+						DataBuffer buffer = this.bufferFactory.allocateBuffer(chunk.buffer.capacity());
+						buffer.write(chunk.buffer);
+						chunk.callback.succeeded();
+						return buffer;
+					});
+
+					return Mono.just(new JettyClientHttpResponse(reactiveResponse, content));
+			}));
+		return Mono.empty();
+	}
+
+	public Mono<ClientHttpResponse> getResponse() {
+		return this.response;
+	}
```

Comment:
It's not obvious at first why a request object should have a method that returns the response, and a public method for that. It must be called at the right time too, or else the response field may not have been initialized, but I can see why the method is here, based on the way the Jetty API works.

I think it would help clarity if the method called `completes` simply stored `ReactiveRequest`, returned `Mono.empty()`, and nothing more. The method already returns `Mono.empty()` anyway, so the part about creating the response is completely independent and could be moved out of the request. In other words, the connector would access the `ReactiveRequest` (via package private method) and then create the response. Then would make it a little more instantly obvious what's going on.

The other point is that the `Mono<Void>` from writeWith is supposed to reflect the completion of the writing of the request body, per contract, but probably not a practical issue at this time. If we keep all this package private we have the option to evolve it in the future.